plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -2249,32 +2248,123 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
+
+        if (columnProjections.values().stream().allMatch(ProjectedColumnRepresentation::isVariable)) {
+            Set<ColumnHandle> projectedColumns = assignments.values().stream()
+                    .map(DeltaLakeColumnHandle.class::cast)
+                    .collect(toImmutableSet());
+            if (deltaLakeTableHandle.getProjectedColumns().isPresent()
+                    && deltaLakeTableHandle.getProjectedColumns().get().equals(projectedColumns)) {
+                return Optional.empty();
+            }
+
+            List<Assignment> newColumnAssignments = assignments.entrySet().stream()
+                    .map(assignment -> new Assignment(
+                            assignment.getKey(),
+                            assignment.getValue(),
+                            ((DeltaLakeColumnHandle) assignment.getValue()).getBaseType()))
+                    .collect(toImmutableList());
+
+            return Optional.of(new ProjectionApplicationResult<>(
+                    deltaLakeTableHandle.withProjectedColumns(projectedColumns),
+                    projections,
+                    newColumnAssignments,
+                    false));
         }
 
-        List<ConnectorExpression> simpleProjections = projections.stream()
-                .filter(projection -> projection instanceof Variable)
-                .collect(toImmutableList());
+        Map<String, Assignment> newAssignments = new HashMap<>();
+        ImmutableMap.Builder<ConnectorExpression, Variable> newVariablesBuilder = ImmutableMap.builder();
+        ImmutableSet.Builder<ColumnHandle> projectedColumnsBuilder = ImmutableSet.builder();
+
+        for (Map.Entry<ConnectorExpression, ProjectedColumnRepresentation> entry : columnProjections.entrySet()) {
+            ConnectorExpression expression = entry.getKey();
+            ProjectedColumnRepresentation projectedColumn = entry.getValue();
+
+            DeltaLakeColumnHandle baseColumnHandle = (DeltaLakeColumnHandle) assignments.get(projectedColumn.getVariable().getName());
+            DeltaLakeColumnHandle projectedColumnHandle = createProjectedColumnHandle(baseColumnHandle, projectedColumn.getDereferenceIndices(), expression.getType());
 
-        List<Assignment> newColumnAssignments = assignments.entrySet().stream()
-                .map(assignment -> new Assignment(
-                        assignment.getKey(),
-                        assignment.getValue(),
-                        ((DeltaLakeColumnHandle) assignment.getValue()).getType()))
+            String projectedColumnName = projectedColumnHandle.getQualifiedName();
+
+            Variable projectedColumnVariable = new Variable(projectedColumnName, expression.getType());
+            Assignment newAssignment = new Assignment(projectedColumnName, projectedColumnHandle, expression.getType());
+            newAssignments.putIfAbsent(projectedColumnName, newAssignment);
```

Comment:
> in this context, if // See if input already contains a columnhandle for this projected column, avoid creating duplicates. comment in HiveMetadata relevant?

Yes, This is relevant, We can use this to avoid creating the same column handle.

Reply:
thanks for confirming the relevance. Should we have similar logic in Delta too?

Reply:
I and @findinpath had an offline discussion and did some experiments with selecting fields with dereference enabled.

> this is redundant (may result in data read twice?)

this is not redundant. Selecting the top-level field along with the leaf does not increase the physical input size.

```
create table mytable1 (id INT, a ROW(b BIGINT, c BIGINT));
```
```
explain analyze select a from mytable1;
--  Input: 10000 rows (224.61kB), Physical input: 127.74kB, Physical input time: 226170000.00ns

explain analyze select a, a.c from mytable1; -- Physical input: 127.74kB
--  Input: 10000 rows (312.50kB), Physical input: 127.74kB, Physical input time: 195990000.00ns

explain analyze select a, a.b, a.c from mytable1; -- Physical input: 127.74kB
--  Input: 10000 rows (400.39kB), Physical input: 127.74kB, Physical input time: 201920000.00ns
```

I will add the test covering this scenario.